File: cortex/python/app/cache/redis.py
"""
Redis Cache Module

This module provides Redis-based caching functionality with support for
namespaces, serialization, and TTL (time-to-live) for cached items.
"""

import json
import logging
from datetime import timedelta
from typing import Any, Dict, List, Optional, Set, TypeVar, Union, cast

import redis.asyncio as redis
from redis.asyncio.client import Redis
from redis.exceptions import RedisError

from app.config import settings
from app.utils.logger import get_contextual_logger

# Configure logger
logger = get_contextual_logger("cache.redis")

# Type variable for generic cache values
T = TypeVar("T")


class RedisCache:
    """
    Redis Cache

    Provides asynchronous Redis caching with namespacing,
    serialization, and TTL management. Designed to work with
    both standalone Redis and Redis clusters.
    """

    def __init__(self, url: Optional[str] = None, prefix: str = ""):
        """
        Initialize Redis cache

        Args:
            url: Redis connection URL (redis://host:port/db)
            prefix: Global key prefix for all cache keys
        """
        self.url = url or settings.redis_url
        self.prefix = prefix or settings.redis_prefix
        self.default_ttl = settings.redis_default_ttl
        self._redis: Optional[Redis] = None
        self._initialized = False

    async def initialize(self) -> bool:
        """
        Initialize Redis connection

        Returns:
            True if connection successful, False otherwise
        """
        if self._initialized and self._redis:
            logger.debug("Redis already initialized")
            return True

        if not self.url:
            logger.warning("Redis URL not configured, Redis cache disabled")
            return False

        try:
            logger.info(f"Initializing Redis connection to {self.url}")
            self._redis = redis.from_url(
                self.url,
                encoding="utf-8",
                decode_responses=True,
                health_check_interval=30,
            )

            # Test connection
            await self._redis.ping()

            self._initialized = True
            logger.info("Redis connection established")
            return True

        except (RedisError, ConnectionError, ValueError) as e:
            logger.error(f"Failed to connect to Redis: {str(e)}")
            self._redis = None
            self._initialized = False
            return False

    async def close(self) -> None:
        """Close Redis connection"""
        if self._redis:
            logger.info("Closing Redis connection")
            await self._redis.close()
            self._redis = None
            self._initialized = False

    def _format_key(self, key: str, namespace: Optional[str] = None) -> str:
        """
        Format cache key with prefix and namespace

        Args:
            key: The cache key
            namespace: Optional namespace

        Returns:
            Formatted key string
        """
        if namespace:
            return f"{self.prefix}{namespace}:{key}"
        else:
            return f"{self.prefix}{key}"

    async def _ensure_connection(self) -> bool:
        """
        Ensure Redis connection is active

        Returns:
            True if connection active, False otherwise
        """
        if not self._initialized or not self._redis:
            return await self.initialize()

        return True

    async def get(
        self, key: str, namespace: Optional[str] = None, default: Any = None
    ) -> Any:
        """
        Get value from cache

        Args:
            key: Cache key
            namespace: Optional namespace
            default: Default value if not found

        Returns:
            Cached value or default if not found
        """
        if not await self._ensure_connection() or not self._redis:
            return default

        try:
            formatted_key = self._format_key(key, namespace)
            value = await self._redis.get(formatted_key)

            if value is None:
                return default

            # Try to deserialize JSON
            try:
                return json.loads(value)
            except json.JSONDecodeError:
                # Return as is if not JSON
                return value

        except RedisError as e:
            logger.error(f"Redis error getting key {key}: {str(e)}")
            return default

    async def set(
        self,
        key: str,
        value: Any,
        namespace: Optional[str] = None,
        ttl: Optional[int] = None,
    ) -> bool:
        """
        Set value in cache

        Args:
            key: Cache key
            value: Value to cache
            namespace: Optional namespace
            ttl: Time-to-live in seconds

        Returns:
            True if successful, False otherwise
        """
        if not await self._ensure_connection() or not self._redis:
            return False

        try:
            formatted_key = self._format_key(key, namespace)

            # Serialize value if necessary
            if isinstance(value, (dict, list, tuple, set)):
                value = json.dumps(value)

            # Set with TTL
            if ttl is None:
                ttl = self.default_ttl

            await self._redis.set(formatted_key, value, ex=ttl)
            return True

        except RedisError as e:
            logger.error(f"Redis error setting key {key}: {str(e)}")
            return False

    async def delete(self, key: str, namespace: Optional[str] = None) -> bool:
        """
        Delete value from cache

        Args:
            key: Cache key
            namespace: Optional namespace

        Returns:
            True if successful, False otherwise
        """
        if not await self._ensure_connection() or not self._redis:
            return False

        try:
            formatted_key = self._format_key(key, namespace)
            await self._redis.delete(formatted_key)
            return True

        except RedisError as e:
            logger.error(f"Redis error deleting key {key}: {str(e)}")
            return False

    async def exists(self, key: str, namespace: Optional[str] = None) -> bool:
        """
        Check if key exists in cache

        Args:
            key: Cache key
            namespace: Optional namespace

        Returns:
            True if key exists, False otherwise
        """
        if not await self._ensure_connection() or not self._redis:
            return False

        try:
            formatted_key = self._format_key(key, namespace)
            return bool(await self._redis.exists(formatted_key))

        except RedisError as e:
            logger.error(f"Redis error checking key {key}: {str(e)}")
            return False

    async def clear_namespace(self, namespace: str) -> int:
        """
        Clear all keys in a namespace

        Args:
            namespace: Namespace to clear

        Returns:
            Number of keys deleted
        """
        if not await self._ensure_connection() or not self._redis:
            return 0

        try:
            pattern = self._format_key("*", namespace)
            keys = await self._redis.keys(pattern)

            if not keys:
                return 0

            # Delete all keys in namespace
            deleted = await self._redis.delete(*keys)
            logger.info(f"Cleared {deleted} keys from namespace {namespace}")
            return deleted

        except RedisError as e:
            logger.error(f"Redis error clearing namespace {namespace}: {str(e)}")
            return 0

    async def increment(
        self,
        key: str,
        value: int = 1,
        namespace: Optional[str] = None,
        ttl: Optional[int] = None,
    ) -> Optional[int]:
        """
        Increment a counter

        Args:
            key: Cache key
            value: Value to increment by
            namespace: Optional namespace
            ttl: Optional new TTL after increment

        Returns:
            New counter value or None if failed
        """
        if not await self._ensure_connection() or not self._redis:
            return None

        try:
            formatted_key = self._format_key(key, namespace)
            pipe = self._redis.pipeline()

            # Increment and get new value
            new_value = await self._redis.incrby(formatted_key, value)

            # Set TTL if provided
            if ttl is not None:
                await self._redis.expire(formatted_key, ttl)

            return new_value

        except RedisError as e:
            logger.error(f"Redis error incrementing key {key}: {str(e)}")
            return None

    async def get_ttl(self, key: str, namespace: Optional[str] = None) -> Optional[int]:
        """
        Get TTL for key

        Args:
            key: Cache key
            namespace: Optional namespace

        Returns:
            TTL in seconds or None if key doesn't exist
        """
        if not await self._ensure_connection() or not self._redis:
            return None

        try:
            formatted_key = self._format_key(key, namespace)
            ttl = await self._redis.ttl(formatted_key)

            # Redis returns -2 if key doesn't exist, -1 if no TTL
            if ttl == -2:
                return None
            if ttl == -1:
                return 0  # No expiration

            return ttl

        except RedisError as e:
            logger.error(f"Redis error getting TTL for key {key}: {str(e)}")
            return None

    async def set_ttl(
        self, key: str, ttl: int, namespace: Optional[str] = None
    ) -> bool:
        """
        Set TTL for key

        Args:
            key: Cache key
            ttl: TTL in seconds
            namespace: Optional namespace

        Returns:
            True if successful, False otherwise
        """
        if not await self._ensure_connection() or not self._redis:
            return False

        try:
            formatted_key = self._format_key(key, namespace)
            return bool(await self._redis.expire(formatted_key, ttl))

        except RedisError as e:
            logger.error(f"Redis error setting TTL for key {key}: {str(e)}")
            return False

    async def get_many(
        self, keys: List[str], namespace: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Get multiple values from cache

        Args:
            keys: List of cache keys
            namespace: Optional namespace

        Returns:
            Dict of key/value pairs for found keys
        """
        if not await self._ensure_connection() or not self._redis:
            return {}

        if not keys:
            return {}

        try:
            # Format all keys
            formatted_keys = [self._format_key(key, namespace) for key in keys]

            # Get all values
            values = await self._redis.mget(formatted_keys)

            # Build result dict
            result = {}
            for i, value in enumerate(values):
                if value is not None:
                    key = keys[i]

                    # Try to deserialize JSON
                    try:
                        result[key] = json.loads(value)
                    except (json.JSONDecodeError, TypeError):
                        # Return as is if not JSON
                        result[key] = value

            return result

        except RedisError as e:
            logger.error(f"Redis error getting multiple keys: {str(e)}")
            return {}

    async def healthcheck(self) -> bool:
        """
        Perform Redis healthcheck

        Returns:
            True if Redis is healthy, False otherwise
        """
        if not self.url:
            # Redis not configured, so it's "healthy" by being disabled
            return True

        try:
            if not await self._ensure_connection() or not self._redis:
                return False

            # Ping Redis
            await self._redis.ping()
            return True

        except RedisError as e:
            logger.error(f"Redis healthcheck failed: {str(e)}")
            return False


# Create global instance
redis_cache = RedisCache()


# Export public symbols
__all__ = ["RedisCache", "redis_cache"]

File: cortex/node/src/cache/redis.ts
/**
 * Redis cache utility for Cortex Core with in-memory fallback
 */

import Redis from 'ioredis';
import config from '../config';
import { logger } from '../utils/logger';

// Redis client instance
let redisClientInstance: Redis | null = null;

// In-memory cache fallback
const memoryCache = new Map<string, {
  value: string;
  expiry: number | null;
}>();

// Track if we're using memory fallback
let usingMemoryFallback = false;

/**
 * Connect to Redis
 */
export async function connectRedis(): Promise<void> {
  try {
    logger.info('Connecting to Redis...');

    // Create Redis client
    redisClientInstance = new Redis({
      host: config.cache.host,
      port: config.cache.port,
      password: config.cache.password,
      retryStrategy: (times) => {
        // If we've failed a few times, use memory fallback
        if (times >= 2) {
          if (!usingMemoryFallback) {
            logger.warn('Redis connection failed, using in-memory cache fallback');
            usingMemoryFallback = true;
          }
          return null; // stop retrying
        }
        const delay = Math.min(times * 50, 2000);
        return delay;
      }
    });

    // Set up event handlers
    redisClientInstance.on('connect', () => {
      usingMemoryFallback = false;
      logger.info('Redis connection established');
    });

    redisClientInstance.on('error', (error) => {
      logger.error('Redis error', error);
      // On error, switch to memory fallback
      if (!usingMemoryFallback) {
        logger.warn('Switching to in-memory cache fallback');
        usingMemoryFallback = true;
      }
    });

    redisClientInstance.on('reconnecting', () => {
      logger.warn('Reconnecting to Redis...');
    });

    try {
      // Test connection
      await redisClientInstance.ping();
      usingMemoryFallback = false;
    } catch (error) {
      logger.warn('Redis ping failed, using in-memory cache fallback');
      usingMemoryFallback = true;
      // We'll continue with the memory fallback
    }

  } catch (error) {
    logger.error('Failed to connect to Redis', error);
    logger.warn('Using in-memory cache fallback');
    usingMemoryFallback = true;
  }

  // Start memory cache cleanup interval
  startMemoryCacheCleanup();
}

/**
 * Disconnect from Redis
 */
export async function disconnectRedis(): Promise<void> {
  try {
    if (redisClientInstance && !usingMemoryFallback) {
      logger.info('Disconnecting from Redis...');
      await redisClientInstance.quit();
      redisClientInstance = null;
      logger.info('Redis connection closed');
    }
  } catch (error) {
    logger.error('Failed to disconnect from Redis', error);
  }
}

/**
 * Start periodic cleanup of expired items in memory cache
 */
function startMemoryCacheCleanup(): void {
  const CLEANUP_INTERVAL = 60 * 1000; // 1 minute

  setInterval(() => {
    const now = Date.now();
    let expiredCount = 0;

    // Clean up expired items
    for (const [key, entry] of memoryCache.entries()) {
      if (entry.expiry !== null && entry.expiry < now) {
        memoryCache.delete(key);
        expiredCount++;
      }
    }

    if (expiredCount > 0) {
      logger.debug(`Memory cache cleanup: removed ${expiredCount} expired items. Current size: ${memoryCache.size}`);
    }
  }, CLEANUP_INTERVAL);
}

/**
 * Redis client wrapper with typed methods that falls back to in-memory storage
 */
export const redisClient = {
  /**
   * Set a key-value pair in Redis or memory
   * @param key The key
   * @param value The value
   * @param expiryMode Optional expiry mode ('EX' for seconds, 'PX' for milliseconds)
   * @param time Optional time value for expiry
   */
  async set(key: string, value: string, expiryMode?: string, time?: number): Promise<string> {
    if (usingMemoryFallback) {
      // Store in memory with expiry
      let expiry: number | null = null;

      if (expiryMode && time) {
        const multiplier = expiryMode === 'EX' ? 1000 : 1; // Convert seconds to ms if needed
        expiry = Date.now() + (time * multiplier);
      }

      memoryCache.set(key, { value, expiry });
      return 'OK';
    } else {
      // Use Redis
      const client = getClient();
      if (expiryMode && time) {
        // Use specific overloads based on the expiry mode
        if (expiryMode === 'EX') {
          return client.set(key, value, 'EX', time);
        } else if (expiryMode === 'PX') {
          return client.set(key, value, 'PX', time);
        } else {
          // Fall back to generic set
          return client.set(key, value);
        }
      } else {
        return client.set(key, value);
      }
    }
  },

  /**
   * Get a value from Redis or memory
   * @param key The key
   * @returns The value or null if key doesn't exist
   */
  async get(key: string): Promise<string | null> {
    if (usingMemoryFallback) {
      const entry = memoryCache.get(key);

      // Return null if key doesn't exist or is expired
      if (!entry || (entry.expiry !== null && entry.expiry < Date.now())) {
        if (entry) {
          // Clean up expired entry
          memoryCache.delete(key);
        }
        return null;
      }

      return entry.value;
    } else {
      const client = getClient();
      return client.get(key);
    }
  },

  /**
   * Delete a key from Redis or memory
   * @param key The key
   * @returns Number of keys removed
   */
  async del(key: string): Promise<number> {
    if (usingMemoryFallback) {
      const existed = memoryCache.delete(key);
      return existed ? 1 : 0;
    } else {
      const client = getClient();
      return client.del(key);
    }
  },

  /**
   * Check if a key exists
   * @param key The key
   * @returns 1 if key exists, 0 otherwise
   */
  async exists(key: string): Promise<number> {
    if (usingMemoryFallback) {
      const entry = memoryCache.get(key);

      // Check if key exists and is not expired
      if (entry && (entry.expiry === null || entry.expiry >= Date.now())) {
        return 1;
      }

      // Clean up expired entry if needed
      if (entry && entry.expiry !== null && entry.expiry < Date.now()) {
        memoryCache.delete(key);
      }

      return 0;
    } else {
      const client = getClient();
      return client.exists(key);
    }
  },

  /**
   * Set expiry time on a key
   * @param key The key
   * @param seconds Expiry time in seconds
   * @returns 1 if timeout was set, 0 if key doesn't exist
   */
  async expire(key: string, seconds: number): Promise<number> {
    if (usingMemoryFallback) {
      const entry = memoryCache.get(key);

      // Return 0 if key doesn't exist or is already expired
      if (!entry || (entry.expiry !== null && entry.expiry < Date.now())) {
        if (entry) {
          // Clean up expired entry
          memoryCache.delete(key);
        }
        return 0;
      }

      // Set new expiry
      entry.expiry = Date.now() + (seconds * 1000);
      memoryCache.set(key, entry);

      return 1;
    } else {
      const client = getClient();
      return client.expire(key, seconds);
    }
  },

  /**
   * Get TTL for a key
   * @param key The key
   * @returns TTL in seconds, -1 if key has no TTL, -2 if key doesn't exist
   */
  async ttl(key: string): Promise<number> {
    if (usingMemoryFallback) {
      const entry = memoryCache.get(key);

      // Return -2 if key doesn't exist
      if (!entry) {
        return -2;
      }

      // Return -1 if key has no expiry
      if (entry.expiry === null) {
        return -1;
      }

      // Check if already expired
      const now = Date.now();
      if (entry.expiry < now) {
        // Clean up expired entry
        memoryCache.delete(key);
        return -2;
      }

      // Return TTL in seconds
      return Math.ceil((entry.expiry - now) / 1000);
    } else {
      const client = getClient();
      return client.ttl(key);
    }
  },

  /**
   * Increment a number stored at key
   * @param key The key
   * @returns The value after incrementing
   */
  async incr(key: string): Promise<number> {
    if (usingMemoryFallback) {
      const entry = memoryCache.get(key);

      // If key doesn't exist, set to 1
      if (!entry) {
        memoryCache.set(key, { value: '1', expiry: null });
        return 1;
      }

      // Parse current value
      const currentValue = parseInt(entry.value, 10) || 0;
      const newValue = currentValue + 1;

      // Update value
      entry.value = newValue.toString();
      memoryCache.set(key, entry);

      return newValue;
    } else {
      const client = getClient();
      return client.incr(key);
    }
  },

  /**
   * Increment a number stored at key by a specified amount
   * @param key The key
   * @param increment The increment amount
   * @returns The value after incrementing
   */
  async incrBy(key: string, increment: number): Promise<number> {
    if (usingMemoryFallback) {
      const entry = memoryCache.get(key);

      // If key doesn't exist, set to increment value
      if (!entry) {
        memoryCache.set(key, { value: increment.toString(), expiry: null });
        return increment;
      }

      // Parse current value
      const currentValue = parseInt(entry.value, 10) || 0;
      const newValue = currentValue + increment;

      // Update value
      entry.value = newValue.toString();
      memoryCache.set(key, entry);

      return newValue;
    } else {
      const client = getClient();
      return client.incrby(key, increment);
    }
  },

  /**
   * Set a key's time to live in milliseconds
   * @param key The key
   * @param milliseconds Expiry time in milliseconds
   * @returns 1 if timeout was set, 0 if key doesn't exist
   */
  async pExpire(key: string, milliseconds: number): Promise<number> {
    if (usingMemoryFallback) {
      const entry = memoryCache.get(key);

      // Return 0 if key doesn't exist or is already expired
      if (!entry || (entry.expiry !== null && entry.expiry < Date.now())) {
        if (entry) {
          // Clean up expired entry
          memoryCache.delete(key);
        }
        return 0;
      }

      // Set new expiry
      entry.expiry = Date.now() + milliseconds;
      memoryCache.set(key, entry);

      return 1;
    } else {
      const client = getClient();
      return client.pexpire(key, milliseconds);
    }
  },

  /**
   * Set key to hold string value if key does not exist
   * @param key The key
   * @param value The value
   * @returns 1 if set, 0 if not set
   */
  async setNX(key: string, value: string): Promise<number> {
    if (usingMemoryFallback) {
      // Check if key exists and is not expired
      const entry = memoryCache.get(key);
      if (entry && (entry.expiry === null || entry.expiry >= Date.now())) {
        return 0; // Key exists, not set
      }

      // Set the value (key doesn't exist or is expired)
      memoryCache.set(key, { value, expiry: null });
      return 1;
    } else {
      const client = getClient();
      return client.setnx(key, value);
    }
  },

  /**
   * Execute a raw Redis command - limited support in memory fallback mode
   * @param command Command name
   * @param args Command arguments
   * @returns Command result
   */
  async exec(command: string, ...args: any[]): Promise<any> {
    if (usingMemoryFallback) {
      // Only support a few basic commands in memory mode
      const lowerCommand = command.toLowerCase();

      switch (lowerCommand) {
        case 'get':
          return this.get(args[0]);
        case 'set':
          return this.set(args[0], args[1]);
        case 'del':
          return this.del(args[0]);
        case 'exists':
          return this.exists(args[0]);
        case 'expire':
          return this.expire(args[0], args[1]);
        case 'ttl':
          return this.ttl(args[0]);
        case 'incr':
          return this.incr(args[0]);
        case 'incrby':
          return this.incrBy(args[0], args[1]);
        default:
          throw new Error(`Command '${command}' not supported in memory fallback mode`);
      }
    } else {
      const client = getClient();
      return client.call(command, ...args);
    }
  },

  /**
   * Flush all data - useful for testing
   */
  async flushAll(): Promise<string> {
    if (usingMemoryFallback) {
      memoryCache.clear();
      return 'OK';
    } else {
      const client = getClient();
      return client.flushall();
    }
  },

  /**
   * Returns true if we're currently using the memory fallback
   */
  isUsingMemoryFallback(): boolean {
    return usingMemoryFallback;
  }
};

/**
 * Get Redis client instance
 * @throws Error if Redis client is not initialized and we're not using memory fallback
 */
function getClient(): Redis {
  if (!redisClientInstance && !usingMemoryFallback) {
    throw new Error('Redis client not initialized');
  }
  return redisClientInstance as Redis;
}

// Export for use in other modules
export default redisClient;
